src/librbd/image/CloneRequest.cc

```diff
@@ -0,0 +1,513 @@
+  if (*r < 0 && *r != -EOPNOTSUPP && *r != -EIO) {
+    lderr(m_cct) << "couldn't list metadata: " << cpp_strerror(*r) << dendl;
+    m_r_saved = *r;
+    send_remove_child();
+  } else if (*r == 0 && !m_pairs.empty()) {
+    send_metadata_set();
+  } else {
+    send_close();
+  }
+
+  return nullptr;
+}
+
+template <typename I>
+void CloneRequest<I>::send_metadata_set() {
+  ldout(m_cct, 20) << this << " " << __func__ << dendl;
+
+  librados::ObjectWriteOperation op;
+  cls_client::metadata_set(&op, m_pairs);
+
+  using klass = CloneRequest<I>;
+  librados::AioCompletion *comp =
+    create_rados_ack_callback<klass, &klass::handle_metadata_set>(this);
+  int r = m_ioctx.aio_operate(m_imctx->header_oid, comp, &op);
+  assert(r == 0);
+  comp->release();
+}
+
+template <typename I>
+Context *CloneRequest<I>::handle_metadata_set(int *r) {
+  ldout(m_cct, 20) << this << " " << __func__ << " r=" << *r << dendl;
+
+  if (*r < 0) {
+    lderr(m_cct) << "couldn't set metadata: " << cpp_strerror(*r) << dendl;
+    m_r_saved = *r;
+    send_remove_child();
+  } else {
+    send_close();
+  }
+
+  return nullptr;
+}
+
+template <typename I>
+void CloneRequest<I>::send_close() {
+  ldout(m_cct, 20) << this << " " << __func__ << dendl;
+
+  assert(m_imctx != nullptr);
+
+  using klass = CloneRequest<I>;
+  Context *ctx = create_async_context_callback(
+    *m_imctx, create_context_callback<
+      klass, &klass::handle_close, false>(this));
+  CloseRequest<I> *req = CloseRequest<I>::create(m_imctx, ctx);
+  req->send();
+}
+
+template <typename I>
+Context *CloneRequest<I>::handle_close(int *r) {
+  ldout(m_cct, 20) << this << " " << __func__ << dendl;
+
+  if (*r < 0) {
+    lderr(m_cct) << "couldn't close image: " << cpp_strerror(*r) << dendl;
+    return m_on_finish;
+  }
+
+  if (m_r_saved == 0) {
+    complete(0);
+  } else {
+    send_remove();
+  }
+  return nullptr;
+}
+
+template <typename I>
+void CloneRequest<I>::send_remove_child() {
+  ldout(m_cct, 20) << this << " " << __func__ << dendl;
+
+  librados::ObjectWriteOperation op;
+  cls_client::remove_child(&op, m_pspec, m_imctx->id);
+
+  using klass = CloneRequest<I>;
+  librados::AioCompletion *comp =
+    create_rados_ack_callback<klass, &klass::handle_remove_child>(this);
+  int r = m_p_imctx->md_ctx.aio_operate(RBD_CHILDREN, comp, &op);
+  assert(r == 0);
+  comp->release();
+}
+
+template <typename I>
+Context *CloneRequest<I>::handle_remove_child(int *r) {
+  ldout(m_cct, 20) << this << " " << __func__ << " r=" << *r << dendl;
+
+  if (*r < 0) {
+     lderr(m_cct) << "Error removing failed clone from list of children: "
+		  << cpp_strerror(*r) << dendl;
+  }
+
+  send_close();
+  return nullptr;
+}
+
+template <typename I>
+void CloneRequest<I>::send_remove() {
+  ldout(m_cct, 20) << this << " " << __func__ << dendl;
+
+  Context *ctx = new FunctionContext([this](int r) {
+    librbd::NoOpProgressContext no_op;
+    r = librbd::remove(m_ioctx, m_name, "", no_op);
+    handle_remove(r);
```

Comment:
yes, that's true. And @vshankar has send a pr for the async remove state machine. https://github.com/ceph/ceph/pull/10896

Maybe I should wait for that pr merged, but before that, am I understanding it correctly? I have to create another thread for the librbd::remove(), instead of calling it in tp_librbd? am I right? thanx @dillaman